src/cmdstan/command_helper.hpp

```diff
@@ -635,6 +675,157 @@ void services_log_prob_grad(const stan::model::model_base &model, bool jacobian,
+  if (num_chains > 1)
+    validate_multi_chain_config(user_method);
+  return num_chains;
+}
+
+/**
+ * Check possible name conflicts between input and output files where both
+ * are in Stan CSV format; if conflicts detected, quit with an error message.
+ *
+ * @param parser user config
+ */
+void check_file_config(argument_parser &parser) {
+  std::string sample_file
+      = get_arg_val<string_argument>(parser, "output", "file");
+  auto user_method = parser.arg("method");
+  if (user_method->arg("generate_quantities")) {
+    std::string input_file = get_arg_val<string_argument>(
+        parser, "method", "generate_quantities", "fitted_params");
+    if (input_file.empty()) {
+      throw std::invalid_argument(
+          std::string("Argument fitted_params file - found empty string, "
+                      "expecting filename."));
+      if (input_file.compare(sample_file) == 0) {
+        std::stringstream msg;
+        msg << "Filename conflict, fitted_params file " << input_file
+            << " and output file names are identical, must be different."
+            << std::endl;
+        throw std::invalid_argument(msg.str());
+      }
+    }
+  } else if (user_method->arg("laplace")) {
+    std::string input_file
+        = get_arg_val<string_argument>(parser, "method", "laplace", "mode");
+    if (input_file.empty()) {
+      throw std::invalid_argument(
+          std::string("Argument mode file - found empty string, "
+                      "expecting filename."));
+      if (input_file.compare(sample_file) == 0) {
+        std::stringstream msg;
+        msg << "Filename conflict, parameter modes file " << input_file
+            << " and output file names are identical, must be different."
+            << std::endl;
+        throw std::invalid_argument(msg.str());
+      }
+    }
+  }
+}
+
+std::vector<std::string> make_filenames(const std::string &filename,
+                                        const std::string &type,
+                                        unsigned int num_chains,
+                                        unsigned int id) {
+  std::vector<std::string> names(num_chains);
+  auto base_sfx = get_basename_suffix(filename);
+  if (base_sfx.second.empty()) {
+    base_sfx.second = type;
+  }
+  auto name_iterator = [num_chains, id](auto i) {
+    if (num_chains == 1) {
+      return std::string("");
+    } else {
+      return std::string("_" + std::to_string(i + id));
+    }
+  };
+  for (int i = 0; i < num_chains; ++i) {
+    names[i] = base_sfx.first + name_iterator(i) + base_sfx.second;
+  }
+  return names;
+}
+
+void init_callbacks(
+    argument_parser &parser,
+    std::vector<stan::callbacks::unique_stream_writer<std::ofstream>>
+        &sample_writers,
+    std::vector<stan::callbacks::unique_stream_writer<std::ofstream>>
+        &diag_csv_writers,
+    std::vector<stan::callbacks::json_writer<std::ofstream>>
+        &diag_json_writers) {
```

Comment:
(optional) I think it could be nice to have a little struct for these that could be like

```c++
struct Callbacks {
  std::vector<stan::callbacks::unique_stream_writer<std::ofstream>> sample_csv;
  std::vector<stan::callbacks::unique_stream_writer<std::ofstream>> diagnostic_csv;
  std::vector<stan::callbacks::json_writer<std::ofstream>> diagnostic_json;
  explicit Callbacks(argument_parser& parser) {
    auto user_method = parser.arg("method");
    unsigned int num_chains = get_num_chains(parser);
    unsigned int id = get_arg_val<int_argument>(parser, "id");
    int sig_figs = get_arg_val<int_argument>(parser, "output", "sig_figs");

    sample_writers.reserve(num_chains);
    std::vector<std::string> output_filenames
        = make_filenames(get_arg_val<string_argument>(parser, "output", "file"),
                        ".csv", num_chains, id);
    for (int i = 0; i < num_chains; ++i) {
      auto ofs = std::make_unique<std::ofstream>(output_filenames[i]);
      if (sig_figs > -1)
        ofs->precision(sig_figs);
      sample_writers.emplace_back(std::move(ofs), "# ");
    }

    diag_json_writers.reserve(num_chains);
    diag_csv_writers.reserve(num_chains);
    // create no-op writers by default
    for (int i = 0; i < num_chains; ++i) {
      diag_json_writers.emplace_back(
          stan::callbacks::json_writer<std::ofstream>());
    }
    for (int i = 0; i < num_chains; ++i) {
      diag_csv_writers.emplace_back(nullptr, "# ");
    }
    // create json, csv writers as needed.
    std::string diagnostic_file
        = get_arg_val<string_argument>(parser, "output", "diagnostic_file");
    if (!diagnostic_file.empty()) {
      std::vector<std::string> diag_filenames;
      if (user_method->arg("pathfinder")) {
        diag_json_writers.clear();
        diag_filenames = make_filenames(diagnostic_file, ".json", num_chains, id);
        for (int i = 0; i < num_chains; ++i) {
          auto ofs = std::make_unique<std::ofstream>(diag_filenames[i]);
          if (sig_figs > -1)
            ofs->precision(sig_figs);
          stan::callbacks::json_writer<std::ofstream> jwriter(std::move(ofs));
          diag_json_writers.emplace_back(std::move(jwriter));
        }
      } else {
        diag_csv_writers.clear();
        diag_filenames = make_filenames(diagnostic_file, ".csv", num_chains, id);
        for (int i = 0; i < num_chains; ++i) {
          auto ofs = std::make_unique<std::ofstream>(diag_filenames[i]);
          if (sig_figs > -1)
            ofs->precision(sig_figs);
          diag_csv_writers.emplace_back(std::move(ofs), "# ");
        }
      }
    }
  }
};
````

Because then we call set this up like 

```c++
Callbacks writers(parser);
```